File: solutions/java/src/parkinglot/ParkingSpot.java
package parkinglot;

import parkinglot.vehicletype.Vehicle;
import parkinglot.vehicletype.VehicleType;

public class ParkingSpot {
    private final int spotNumber;
    private final VehicleType vehicleType;
    private Vehicle vehicle;
    private boolean isOccupied;

    /**
     * Exp PL 01:
     * Objects state is comprised of all of the above variables.
     * So any access to these variables should be synchronized.
     */

    private Object lock = new Object();

    public ParkingSpot(int spotNumber, VehicleType vehicleType) {
        this.spotNumber = spotNumber;
        this.vehicleType = vehicleType;
        this.isOccupied = false;
    }

    public boolean isAvailable() {
        // sync on primitive(boolean) does not work. Hence use lock variable
        // Ideally i would have wanted to use isOccupied variable directly.
        // synchronized(this.isOccupied)
        // {
        // return !isOccupied;
        // }

        synchronized (lock) {
            return !isOccupied;
        }

    }

    public boolean park(Vehicle vehicle) {
        synchronized (lock) {
            // Check if the spot is already occupied or if the vehicle type does not match
            if (isOccupied || vehicle.getType() != vehicleType) {
                return false;
            }
            this.vehicle = vehicle;
            isOccupied = true;
        }
        return true;
    }

    public void unpark() {
        /**
         * See Exp PL 01
         */
        synchronized (lock) {
            vehicle = null;
            isOccupied = false;
        }

    }

    public VehicleType getVehicleType() {
        /**
         * See Exp PL 01
         */
        synchronized (lock) {
            return vehicleType;
        }
    }

    public Vehicle getVehicle() {
        /**
         * See Exp PL 01
         */
        synchronized(lock)
        {
            return vehicle;
        }
    }

    public int getSpotNumber() {
        /**
         * See Exp PL 01
         */
        synchronized (lock) {
            return spotNumber;
        }   
    }
}

File: solutions/cpp/parkinglot/ParkingSpot.cpp
#include "ParkingSpot.hpp"
#include <iostream>

ParkingSpot::ParkingSpot(int spotNumber, SpotType type)
    : spotNumber(spotNumber), type(type), vehicle(nullptr), available(true) {}

int ParkingSpot::getSpotNumber() const { return spotNumber; }
SpotType ParkingSpot::getType() const { return type; }
Vehicle* ParkingSpot::getVehicle() const { return vehicle; }
bool ParkingSpot::isAvailable() const { return available; }

bool ParkingSpot::canFitVehicle(const Vehicle* vehicle) const {
    if (!vehicle) return false;
    
    switch (vehicle->getType()) {
        case VehicleType::MOTORCYCLE:
            return true;  // Can park in any spot
            
        case VehicleType::CAR:
            return type != SpotType::COMPACT;
            
        case VehicleType::TRUCK:
        case VehicleType::BUS:
            return type == SpotType::LARGE;
    }
    return false;
}

bool ParkingSpot::parkVehicle(Vehicle* vehicle) {
    if (!available || !canFitVehicle(vehicle)) return false;
    
    this->vehicle = vehicle;
    available = false;
    return true;
}

Vehicle* ParkingSpot::removeVehicle() {
    if (!vehicle) return nullptr;
    
    Vehicle* removedVehicle = vehicle;
    vehicle = nullptr;
    available = true;
    return removedVehicle;
}

void ParkingSpot::displayInfo() const {
    std::cout << "Spot " << spotNumber << " (";
    switch (type) {
        case SpotType::COMPACT: std::cout << "Compact"; break;
        case SpotType::REGULAR: std::cout << "Regular"; break;
        case SpotType::LARGE: std::cout << "Large"; break;
    }
    std::cout << "): " << (available ? "Available" : "Occupied");
    if (vehicle) {
        std::cout << " by ";
        vehicle->displayInfo();
    } else {
        std::cout << std::endl;
    }
} 